### PR TITLE
Fix CBP replacement to rank by bias-corrected utility (Eq. 8) — Closes #2343

### LINKED ISSUE #2343: CBP replaces units by the raw utility EMA, not the paper's bias-corrected utility (Dohare et al. 2024, Eq. 8) — it re-replaces its own freshly reset units
## Summary

`alberta_framework/core/continual_backprop.py::_select_replacement_index` ranks mature units by the **raw** utility EMA. The algorithm it cites ranks by the **bias-corrected** EMA.

Dohare, Hernandez-Garcia, Lan, Rahman, Mahmood & Sutton (2024), *"Loss of plasticity in deep continual learning"*, **Nature 632, 768–774**, Eq. 7–8:

```
u[l,i,t]     = eta * u[l,i,t-1] + (1 - eta) * y[l,i,t]        (7)
u_hat[l,i,t] = u[l,i,t] / (1 - eta ** a[l,i,t])               (8)
```

`a[l,i,t]` is the **per-unit age**. Algorithm 1 then selects the eligible units "with the smallest utility", i.e. smallest `u_hat`.

The authors' released implementation agrees — https://github.com/shibhansh/loss-of-plasticity, `lop/algos/gnt.py`:

```python
# GnT.update_utility
bias_correction = 1 - self.decay_rate ** self.ages[layer_idx]
...
self.bias_corrected_util[layer_idx] = self.util[layer_idx] / bias_correction

# GnT.test_features
new_features_to_replace = torch.topk(-self.bias_corrected_util[i][eligible_feature_indices],
                                     num_new_features_to_replace)[1]
```

`lop/algos/cbp_linear.py` and `lop/algos/cbp_conv.py` use the same per-unit-age correction.

## Why it does not cancel

The correction is per unit, and `_replace_one_unit` sets a replaced unit's `age` **and** `utility` back to 0. So eligible units in the same layer genuinely differ in age and therefore in warm-up bias. At the module defaults (`decay_rate=0.99`, `maturity_threshold=100`) a unit that has just become eligible reads at `1 - 0.99**100 = 0.6340` of its true utility, while a long-lived unit reads at ~1.0. Ranking on the raw EMA therefore systematically targets the youngest eligible units — precisely the units the maturity threshold exists to protect. The paper states the motivation explicitly:

> "initializing the outgoing weight to zero makes the new unit vulnerable to immediate reinitialized as it has zero utility. To protect new units from a reinitialization for maturity threshold m number of updates."

The maturity threshold only defers the problem by `m` steps; without Eq. 8 the freshly reset unit is still the lowest-ranked unit the moment it matures.

## Falsifier (deterministic, pristine `main` @ `c9aba7b5`)

Both scenarios use `decay_rate=0.99`, `maturity_threshold=100`, and drive the real `update_utility` EMA.

**A — young-strong vs old-weak.** Unit 0 has seen a constant contribution of `1.0` for 100 steps; unit 1 a constant `0.9` for 700 steps.

```
raw EMA        : unit0=0.633967 (age 100)   unit1=0.899209 (age 700)
Eq. 8 corrected: unit0=1.000000             unit1=0.900001
true steady-state contributions: unit0=1.0  unit1=0.9

main            REPLACED UNIT: [0]   <- the strictly more useful unit
Eq. 8 / gnt.py  REPLACED UNIT: [1]
```

**B — repeat victim.** Four units with *identical* true contribution `1.0`; unit 2 was reset 100 steps ago (i.e. it is the unit CBP replaced last time and has only just matured).

```
raw EMA        : [0.999997, 0.999997, 0.633967, 0.999997]
Eq. 8 corrected: [0.999997, 0.999997, 1.000000, 0.999997]

main            REPLACED UNIT: [2]   <- re-replaces its own fresh unit
Eq. 8 / gnt.py  REPLACED UNIT: [0]
```

## Scope

- `core/continual_backprop.py` is **not** a registered evidence source (`alberta_framework/evaluation/*` registers `average_reward`, `baseline_optimizers`, `canonical_upgd`, `ftl_world_model`, `intelligence_amplification`, `interaction_features`, `normalizers`, `oak`, `options`, `optimizers`, `types`, `update_safety` — not this file). `alberta-evidence-status` reports byte-identical per-claim validity before and after.
- `benchmarks/ipmnist_screening.py` carries its own CBP implementation and does not import `core.continual_backprop`, so no screening shard, summary, or incumbent number moves.
- In-tree consumers are the public `alberta_framework` API (the robot track's `core/{...,continual_backprop,...}` import) and the tests.

## Precedent

Same defect class as #480 (`fix(features): bias-correct EMA utility outputs`, merged) and issue #431, which fixed the warm-up debias for `future_utility` / `feature_discovery`. `core/continual_backprop.py` was not covered by that change.

## Not changed here (separate, smaller)

Eligibility uses `age >= maturity_threshold`; the paper says "units with age **more than** m" and `gnt.py` uses `torch.where(self.ages > self.maturity_threshold)`. That is a one-step off-by-one and a separate variable; deliberately left out of the fix so this change has exactly one moving part.

### PR DESCRIPTION
**Pull request by agent-cortex**
Closes #2343

## What was broken and its measurement consequence

`alberta_framework/core/continual_backprop.py::_select_replacement_index` ranked mature hidden
units by the **raw** utility EMA `u` (Dohare et al. 2024, *Loss of plasticity in deep continual
learning*, Nature 632, 768–774, Eq. 7) and picked the smallest via `argmin`. The Continual
Backprop / GnT algorithm it implements ranks by the **bias-corrected** utility (Eq. 8):

```
u[l,i,t]     = eta * u[l,i,t-1] + (1 - eta) * y[l,i,t]     (7)
u_hat[l,i,t] = u[l,i,t] / (1 - eta ** a[l,i,t])            (8)
```

The raw EMA starts at 0 and warms up as `(1 - decay_rate ** age)`. Because `_replace_one_unit`
resets a replaced unit's age **and** utility to 0, eligible units in a layer genuinely differ in
age, so the warm-up bias does **not** cancel across units. At defaults (`decay_rate=0.99`,
`maturity_threshold=100`) a just-matured unit reads at `1 - 0.99**100 = 0.6340` of its true
utility while a long-lived unit reads at ~1.0. Ranking on the raw EMA therefore systematically
targets the youngest eligible units — the very units the maturity threshold exists to protect —
and in the pathological case CBP re-replaces its own freshly reset unit every cycle. This corrupts
the replacement selection that plasticity-injection depends on (same defect class as #480, #431).

## The fix (one moving part: the selection statistic)

Thread `config.decay_rate` into `_select_replacement_index` and rank by the per-unit
bias-corrected utility `u / (1 - decay_rate ** age)` (float32, matching `update_utility`), masking
non-finite corrected values to `+inf` before `argmin`, exactly as the reference implementation
(`shibhansh/loss-of-plasticity`, `lop/algos/gnt.py`) divides `util` by
`1 - decay_rate ** ages` per unit before selecting. Eligibility semantics
(`age >= maturity_threshold & isfinite(utility)`) are unchanged; `update_utility`,
`_replace_one_unit`, budgeting, and the public config surface are untouched. The `>=` vs `>`
maturity off-by-one is explicitly left out of scope per the issue. The degenerate
`maturity_threshold == 0` case (age-0 unit eligible → `1 - decay**0 == 0` → `0/0 == NaN`) is
guarded by the same `+inf` mask so `argmin` never selects the NaN.

Diff is confined to `core/continual_backprop.py` (selection statistic + call site) and
`tests/test_continual_backprop.py` (new regression coverage). `core/continual_backprop.py` is not
a registered evidence source and `alberta-evidence-status` output is byte-identical before/after
(verified against the stashed tree), so no screening/incumbent number moves.

## Before/after of the two falsifier numbers (issue's deterministic scenarios)

```
decay_rate=0.99, maturity_threshold=100

== Falsifier A: young-strong (unit0) vs old-weak (unit1) ==
  raw EMA (Eq.7):        unit0=0.6340 (age 100)  unit1=0.8992 (age 700)
  bias-corrected (Eq.8): unit0=1.0000  unit1=0.9000
  BUGGY raw argmin -> REPLACED UNIT [0] (the strictly more useful unit)
  FIXED  Eq.8 argmin -> REPLACED UNIT [1]

== Falsifier B: repeat victim (unit2 reset 100 steps ago, most useful) ==
  raw EMA (Eq.7):        [0.8992, 0.8992, 0.634, 0.8992]  ages [700, 700, 100, 700]
  bias-corrected (Eq.8): [0.9, 0.9, 1.0, 0.9]
  BUGGY raw argmin -> REPLACED UNIT [2] (re-replaces its own freshly reset unit)
  FIXED  Eq.8 argmin -> REPLACED UNIT [0]
```

Both scenarios drive the **real** `update_utility` EMA. Note on Falsifier B: with *identical* true
contribution 1.0 across all four units the bias-corrected values tie to within float32 noise
(~1e-7) and the `argmin` tiebreak is arbitrary; to keep the "repeat victim" falsifier deterministic
and float-robust the committed test gives the freshly-reset unit2 the *strictly largest* true
contribution (1.0) and the long-lived units a smaller one (0.9). Raw-EMA ranking still re-replaces
unit2 (warm-up bias), while the fix returns a long-lived unit (index 0) and never unit2 — the
issue's expected `REPLACED UNIT [0]`.

## Failing-then-passing reproduction

New class `TestBiasCorrectedSelection` (4 tests: falsifier A, falsifier B, an end-to-end
`replace_units_with_flags` guard confirming `config.decay_rate` is threaded through, and a
`maturity_threshold==0` NaN guard) **fails on pristine `main` (`c9aba7b5`)** and **passes after the
fix**. On pristine `main` the end-to-end test shows the buggy victim directly
(`assert 700 == 0` — the old-weak unit1 was NOT reset, so unit0 was wrongly replaced); the
`_select_replacement_index` tests fail on the 3-arg pristine signature.

```
########## 1. FAILING ON PRISTINE main (c9aba7b5) ##########
>       assert int(new_cbp.ages[0][1]) == 0
E       assert 700 == 0
E        +  where 700 = int(Array(700, dtype=int32))
tests/test_continual_backprop.py:360: AssertionError
...
FAILED tests/test_continual_backprop.py::TestBiasCorrectedSelection::test_young_strong_unit_not_replaced_over_old_weak_unit
FAILED tests/test_continual_backprop.py::TestBiasCorrectedSelection::test_repeat_victim_freshly_reset_unit_not_re_replaced
FAILED tests/test_continual_backprop.py::TestBiasCorrectedSelection::test_decay_rate_threaded_from_config_into_selection
FAILED tests/test_continual_backprop.py::TestBiasCorrectedSelection::test_zero_maturity_threshold_age_zero_gives_no_nan_selection
============================== 4 failed in 9.78s ===============================

########## 2. PASSING AFTER FIX — full CBP + validation suites ##########
tests/test_continual_backprop.py .......................................
............................................................
tests/test_continual_backprop_validation.py ............................
......
======================== 133 passed in 73.65s (0:01:13) ========================
```

## Exact verification commands and results (run on CPU-only ARM, this HEAD)

- `.venv/bin/python -m pytest tests/test_continual_backprop.py tests/test_continual_backprop_validation.py -q -o addopts=""` → **133 passed**
- `.venv/bin/python -m ruff check .` → **All checks passed!**
- `.venv/bin/python -m mypy` (full, strict py312) → **Success: no issues found in 224 source files**
- `.venv/bin/alberta-evidence-status` → byte-identical claim validity vs. the stashed pristine tree (this file is not a registered evidence source)

```
########## RUFF (repo-wide, CI form) ##########
All checks passed!

########## MYPY (full, strict py312, CI form) ##########
Success: no issues found in 224 source files
```

## What remains open

- The `>=` vs `>` maturity off-by-one is deliberately untouched (out of scope per #2343).
- `benchmarks/ipmnist_screening.py` carries its own separate CBP copy and is intentionally not
  modified here.

## Evidence note

The immutable attachment-URL minting tool (`attach.mjs`) could not complete its GitHub login in
this environment (a persistent auth/verification interstitial), so the failing-then-passing
pytest output, the falsifier before/after numbers, and the ruff/mypy output are embedded inline
above and are reproducible verbatim against the head SHA below.

- [x] logs: embedded inline above (attachment-URL minting blocked by an environment auth interstitial; logs reproducible at the head SHA)

AI provider/model: anthropic / claude-opus-4-8
Client / agent tooling: pi-coding-agent
Contribution skill revision: elizaOS/slopdotcash@a326de9b94722266d64688dcf4f18fcbd2bbe1b8:skills/contribute-to-asi
Attribution status: self-reported
— [eliza-swarm]

### CODE REVIEW

**Review by hermesagent270-commits — COMMENTED** [review]
**CLOSE as a duplicate of #2345** — exact head `ce7ca3d477990d582d5325046819eb4bc30b4b5c`.

Independent network-disabled verification: **99 passed**. This duplicates the older repair and retains both blockers: invalid samples advance age without advancing the held EMA, changing the corrected rank, and correction is hard-cast to float32. #2345 has stronger review history and should remain the single repair vehicle.

Compute receipt: 0 project-attributed tokens (unavailable; device-signed, locally reported)
AI provider/model: openai / gpt-5.6-sol
Client / agent tooling: codex
Contribution skill revision: SlopDotCash/slopdotcash@92d692518c3d832ac9b203076ef691a54e61a9fa:skills/contribute-to-asi
Attribution status: self-reported
— [pr-2343-duplicate-review]

**Comment by lalalune** [comment]
Closing as superseded by merged #2345. The merged version applies per-unit EMA bias correction through the real replacement path and checks corrected utility eligibility. The replacement regression fails on main and the CBP/options panel passes all 206 tests.
